File: services/academic-service/src/main/java/com/visor/school/academicservice/model/AcademicRecord.java
package com.visor.school.academicservice.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Academic record representing a student's complete academic history
 */
@Entity
@Table(name = "academic_records", indexes = {
    @Index(name = "idx_academic_records_student", columnList = "student_id", unique = true),
    @Index(name = "idx_academic_records_standing", columnList = "academic_standing")
})
@Data
@NoArgsConstructor
@AllArgsConstructor
public class AcademicRecord {
    private static final BigDecimal MAX_GPA = new BigDecimal("4.0");

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "student_id", nullable = false, unique = true)
    private UUID studentId;

    @ElementCollection
    @CollectionTable(name = "enrollment_history", joinColumns = @JoinColumn(name = "academic_record_id"))
    private List<EnrollmentEntry> enrollmentHistory = new ArrayList<>();

    @ElementCollection
    @CollectionTable(name = "course_completions", joinColumns = @JoinColumn(name = "academic_record_id"))
    private List<CourseCompletion> completedCourses = new ArrayList<>();

    @Column(name = "current_gpa", nullable = false, precision = 3, scale = 2)
    private BigDecimal currentGPA = BigDecimal.ZERO;

    @Column(name = "cumulative_gpa", nullable = false, precision = 3, scale = 2)
    private BigDecimal cumulativeGPA = BigDecimal.ZERO;

    @Column(name = "credits_earned", nullable = false)
    private int creditsEarned = 0;

    @Column(name = "credits_required", nullable = false)
    private int creditsRequired = 120;

    @Enumerated(EnumType.STRING)
    @Column(name = "academic_standing", nullable = false)
    private AcademicStanding academicStanding = AcademicStanding.GOOD_STANDING;

    @Column(name = "graduation_date")
    private LocalDate graduationDate;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt = Instant.now();

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt = Instant.now();

    public AcademicRecord(UUID studentId) {
        this.studentId = studentId;
        validateGPA(BigDecimal.ZERO);
        validateCredits(0, 120);
    }

    public AcademicRecord(UUID studentId, BigDecimal currentGPA, BigDecimal cumulativeGPA, int creditsEarned, AcademicStanding academicStanding) {
        validateGPA(currentGPA);
        validateGPA(cumulativeGPA);
        validateCredits(creditsEarned, 120);
        
        this.studentId = studentId;
        this.currentGPA = currentGPA;
        this.cumulativeGPA = cumulativeGPA;
        this.creditsEarned = creditsEarned;
        this.academicStanding = academicStanding;
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
    }

    public AcademicRecord(UUID studentId, BigDecimal currentGPA, BigDecimal cumulativeGPA, int creditsEarned, int creditsRequired, AcademicStanding academicStanding) {
        validateGPA(currentGPA);
        validateGPA(cumulativeGPA);
        validateCredits(creditsEarned, creditsRequired);
        
        this.studentId = studentId;
        this.currentGPA = currentGPA;
        this.cumulativeGPA = cumulativeGPA;
        this.creditsEarned = creditsEarned;
        this.creditsRequired = creditsRequired;
        this.academicStanding = academicStanding;
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
    }

    private void validateGPA(BigDecimal gpa) {
        if (gpa.compareTo(BigDecimal.ZERO) < 0 || gpa.compareTo(MAX_GPA) > 0) {
            throw new IllegalArgumentException("GPA must be between 0.0 and 4.0, got: " + gpa);
        }
    }

    private void validateCredits(int earned, int required) {
        if (earned < 0) {
            throw new IllegalArgumentException("Credits earned cannot be negative, got: " + earned);
        }
        if (required <= 0) {
            throw new IllegalArgumentException("Credits required must be positive, got: " + required);
        }
    }

    public void updateGPA(BigDecimal current, BigDecimal cumulative) {
        validateGPA(current);
        validateGPA(cumulative);
        this.currentGPA = current;
        this.cumulativeGPA = cumulative;
        this.updatedAt = Instant.now();
    }

    public void updateAcademicStanding(AcademicStanding standing) {
        this.academicStanding = standing;
        this.updatedAt = Instant.now();
    }

    public void markAsGraduated(LocalDate date) {
        this.academicStanding = AcademicStanding.GRADUATED;
        this.graduationDate = date;
        this.updatedAt = Instant.now();
    }

    public void addCompletedCourse(CourseCompletion course) {
        completedCourses.add(course);
        creditsEarned += course.getCredits();
        updatedAt = Instant.now();
    }

    public void addEnrollmentEntry(EnrollmentEntry entry) {
        enrollmentHistory.add(entry);
        updatedAt = Instant.now();
    }
}

File: services/academic-service/src/main/kotlin/com/visor/school/academicservice/model/AcademicRecord.kt
package com.visor.school.academicservice.model

import jakarta.persistence.*
import java.math.BigDecimal
import java.time.Instant
import java.time.LocalDate
import java.util.UUID

/**
 * Academic record representing a student's complete academic history
 */
@Entity
@Table(name = "academic_records", indexes = [
    Index(name = "idx_academic_records_student", columnList = "student_id", unique = true),
    Index(name = "idx_academic_records_standing", columnList = "academic_standing")
])
class AcademicRecord(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: UUID = UUID.randomUUID(),

    @Column(name = "student_id", nullable = false, unique = true)
    val studentId: UUID,

    @ElementCollection
    @CollectionTable(name = "enrollment_history", joinColumns = [JoinColumn(name = "academic_record_id")])
    @Embedded
    val enrollmentHistory: MutableList<EnrollmentEntry> = mutableListOf(),

    @ElementCollection
    @CollectionTable(name = "course_completions", joinColumns = [JoinColumn(name = "academic_record_id")])
    @Embedded
    val completedCourses: MutableList<CourseCompletion> = mutableListOf(),

    @Column(name = "current_gpa", nullable = false, precision = 3, scale = 2)
    var currentGPA: BigDecimal = BigDecimal.ZERO,

    @Column(name = "cumulative_gpa", nullable = false, precision = 3, scale = 2)
    var cumulativeGPA: BigDecimal = BigDecimal.ZERO,

    @Column(name = "credits_earned", nullable = false)
    var creditsEarned: Int = 0,

    @Column(name = "credits_required", nullable = false)
    val creditsRequired: Int = 120,

    @Enumerated(EnumType.STRING)
    @Column(name = "academic_standing", nullable = false)
    var academicStanding: AcademicStanding = AcademicStanding.GOOD_STANDING,

    @Column(name = "graduation_date")
    var graduationDate: LocalDate? = null,

    @Column(name = "created_at", nullable = false)
    val createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false)
    var updatedAt: Instant = Instant.now()
) {
    init {
        require(currentGPA >= BigDecimal.ZERO && currentGPA <= BigDecimal("4.0")) {
            "Current GPA must be between 0.0 and 4.0, got: $currentGPA"
        }
        require(cumulativeGPA >= BigDecimal.ZERO && cumulativeGPA <= BigDecimal("4.0")) {
            "Cumulative GPA must be between 0.0 and 4.0, got: $cumulativeGPA"
        }
        require(creditsEarned >= 0) {
            "Credits earned cannot be negative, got: $creditsEarned"
        }
        require(creditsRequired > 0) {
            "Credits required must be positive, got: $creditsRequired"
        }
    }

    fun updateGPA(current: BigDecimal, cumulative: BigDecimal) {
        require(current >= BigDecimal.ZERO && current <= BigDecimal("4.0")) {
            "Current GPA must be between 0.0 and 4.0, got: $current"
        }
        require(cumulative >= BigDecimal.ZERO && cumulative <= BigDecimal("4.0")) {
            "Cumulative GPA must be between 0.0 and 4.0, got: $cumulative"
        }
        currentGPA = current
        cumulativeGPA = cumulative
        updatedAt = Instant.now()
    }

    fun updateAcademicStanding(standing: AcademicStanding) {
        academicStanding = standing
        updatedAt = Instant.now()
    }

    fun markAsGraduated(date: LocalDate) {
        academicStanding = AcademicStanding.GRADUATED
        graduationDate = date
        updatedAt = Instant.now()
    }

    fun addCompletedCourse(course: CourseCompletion) {
        completedCourses.add(course)
        creditsEarned += course.credits
        updatedAt = Instant.now()
    }

    fun addEnrollmentEntry(entry: EnrollmentEntry) {
        enrollmentHistory.add(entry)
        updatedAt = Instant.now()
    }
}

